src/pytkdocs/parsers/docstrings/__init__.py

```diff
@@ -0,0 +1,122 @@
+import inspect
+from abc import ABCMeta
+from typing import Any, List, Optional
+
+empty = inspect.Signature.empty
+
+
+class AnnotatedObject:
+    """A helper class to store information about an annotated object."""
+
+    def __init__(self, annotation, description):
+        self.annotation = annotation
+        self.description = description
+
+
+class Parameter(AnnotatedObject):
+    """A helper class to store information about a signature parameter."""
+
+    def __init__(self, name, annotation, description, kind, default=empty):
+        super().__init__(annotation, description)
+        self.name = name
+        self.kind = kind
+        self.default = default
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return f"<Parameter({self.name}, {self.annotation}, {self.description}, {self.kind}, {self.default})>"
+
+    @property
+    def is_optional(self):
+        return self.default is not empty
+
+    @property
+    def is_required(self):
+        return not self.is_optional
+
+    @property
+    def is_args(self):
+        return self.kind is inspect.Parameter.VAR_POSITIONAL
+
+    @property
+    def is_kwargs(self):
+        return self.kind is inspect.Parameter.VAR_KEYWORD
+
+    @property
+    def default_string(self):
+        if self.is_kwargs:
+            return "{}"
+        elif self.is_args:
+            return "()"
+        elif self.is_required:
+            return ""
+        return repr(self.default)
+
+
+class Section:
+    """A helper class to store a docstring section."""
+
+    class Type:
+        MARKDOWN = "markdown"
+        PARAMETERS = "parameters"
+        EXCEPTIONS = "exceptions"
+        RETURN = "return"
+
+    def __init__(self, section_type, value):
+        self.type = section_type
+        self.value = value
+
+    def __str__(self):
+        return self.type
+
+    def __repr__(self):
+        return f"<Section(type={self.type!r})>"
+
+
+class Parser(metaclass=ABCMeta):
+    """
+    A class to parse docstrings.
+
+    It is instantiated with an object's path, docstring, signature and return type.
+
+    The `parse` method then returns structured data,
+    in the form of a list of [`Section`][pytkdocs.parsers.docstrings.Section]s.
+    It also return the list of errors that occurred during parsing.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        docstring: str,
```

Comment:
Do we want the `Parser`s defined such that we need a new instance for each docstring? It seems like we're making `__init__` unambiguous and making `do_parse` ambiguous because it might get configuration options. The parser behavior should be relatively stable, so it might be better to make `__init__`  the configuration portion, thus it becomes ambiguous with potential `**kwargs`. Then, the `docstring`, `path`, `signature`, and `return_type` get passed into `do_parse` with nothing else.

Reply:
Thanks, I didn't like it either, your suggestion is nice!

It would roughly look like this:

```
class Parser:
    def __init__(self, **options):
        # set options
        self._errors = []

    def parse(self, docstring, path, signature, return_type):
        self._errors = []  # reset
        return self.do_parse(docstring, path, signature, return_type), self._errors
```

Then the implementation would use `self._errors` to collect errors (and it's private because nobody should access `parser._errors` since it's reset each time we call `parser.parse()`).

I think I'll also invert `parse` and `do_parse`: it makes more sense when implementing a parser to define the `parse` function than a `do_parse` one.

Reply:
You don't even need `**options` in the base `Parser` since it's not being used. It doesn't prevent child implementations from using such a format.

Reply:
Alright, I see. Linting tools usually complains that an implementation method doesn't have the same signature as its parent. I'll play with that in the code :slightly_smiling_face:

Reply:
In fact, 4 attributes are needed when parsing: errors, path, signature and return type. So when calling `do_parse` (still don't like the name though), I reset these private attribute on the parser instance, so the actual `parse` implemented method can use them while parsing.